executor/write.go

```diff
@@ -913,6 +924,136 @@ func (e *InsertExec) exec(goCtx goctx.Context, rows [][]types.Datum) (Row, error
 	return nil, nil
 }
 
+type keyWithDupError struct {
+	key    kv.Key
+	dupErr error
+}
+
+func (e *InsertExec) getRecordIDs(rows [][]types.Datum) ([]int64, error) {
+	recordIDs := make([]int64, 0, len(rows))
+	if e.Table.Meta().PKIsHandle {
+		var handleCol *table.Column
+		for _, col := range e.Table.Cols() {
+			if col.IsPKHandleColumn(e.Table.Meta()) {
+				handleCol = col
+				break
+			}
+		}
+		for _, row := range rows {
+			recordIDs = append(recordIDs, row[handleCol.Offset].GetInt64())
+		}
+	} else {
+		for range rows {
+			recordID, err := e.Table.AllocAutoID(e.ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			recordIDs = append(recordIDs, recordID)
+		}
+	}
+	return recordIDs, nil
+}
+
+// getKeysNeedCheck gets keys converted from to-be-insert rows to record keys and unique index keys,
+// which need to be checked whether they are duplicate keys.
+func (e *InsertExec) getKeysNeedCheck(rows [][]types.Datum) ([][]keyWithDupError, error) {
+	nUnique := 0
+	for _, v := range e.Table.WritableIndices() {
+		if !v.Meta().Unique && !v.Meta().Primary {
```

Comment:
If the index is a primary index, the unique must be true. So I think we can remove `!v.Meta().Primary`. 
The following code maybe enough:
```
if v.Metat().Unique{
    nUnique++
}
```